_posts/2018-10-02-auto-tune-all.md

```diff
@@ -0,0 +1,198 @@
+
+
+### NVIDIA GPU
+
+On NVIDIA GPU, [CuDNN](https://developer.nvidia.com/cudnn) and [TensorRT](https://developer.nvidia.com/tensorrt) are two vendor-provided libraries for training and inference respectively. Since we focus on inference,
+we run our benchmark on unbatched setting. Another tensor compiler [PlaidML](https://github.com/plaidml/plaidml) is also reported as baseline.
+We get the results from its benchmark tool [PlaidBench](https://github.com/plaidml/plaidbench).
+
+According to the results, we reaches the TensorRT level. It is really a strong baseline as we all know how NVIDIA dominates deep learning hardware.
+
+![image](/images/autotune-all/nvidia.png){: width="90%"}
+
+### AMD GPU
+
+Finally let we take a look at AMD GPU. TVM supports OpenCL and [ROCm](https://rocm.github.io/) backend. We found ROCm is better since
+it is more specialized for AMD GPUs. In terms of baseline, [MIOpen](https://github.com/ROCmSoftwarePlatform/MIOpen) is a vendor provided
+kernel library. We integrate its kernel implementation in TVM graph runtime.
+
+We didn't do any specific optimization for AMD GPU. Instead, all computation definition/schedule code for NVIDIA GPU is directly reused. As for the results, TVM is a little bit slower then MIOpen in most cases.
+
+![image](/images/autotune-all/amd.png){: width="90%"}
+
+## All Our Results
+We tested the following networks on ImageNet (3x224x224) dataset with batch size = 1 and data type = float32.
+The reported numbers are time costs per image in millisecond.
+
+| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| |
+| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| | **densenet121** | **inception v3** | **mobilenet** | **mobilenet v2** | **resnet18** | **resnet50** | **squeezenet v1.0** | **squeezenet v1.1** | **vgg16** | **vgg19** |
+| **ARM CPU** |
+| Huawei P20 Pro | 181.4 | 439.9 | 56.1 | 34.5 | 76.5 | 208.2 | 51.8 | 25.7 | 480.6 | 627.0 |
+| Google Pixel2 | 162.2 | 433.5 | 39.5 | 30.1 | 61.1 | 181.3 | 47.3 | 23.2 | 391.1 | 487.7 |
+| Xilinx PYNQ | 2888.3 | 9709.1 | 723.5 | 514.3 | 1234.6 | 3580.5 | 909.9 | 477.3 | <sup>-(Note 1)</sup>  | - |
+| Firefly RK3399 | 335.9 | 1285.9 | 78.6 | 66.7 | 161.2 | 403.8 | 94.6 | 48.5 | 902.9 | 1090.1 |
+| Raspberry Pi 3B | 608.4 | 2065.7 | 122.3 | 103.9 | 314.4 | 722.8 | 185.2 | 94.2 | 1801.4 | 2124.2 |
+| **Mali GPU** |
+| Mali-T860 | 461.5 | 1055.6 | 75.1 | 85.4 | 127.8 | 412.5 | 108.7 | 57.6 | 678.1 | 799.1 |
+| **NVIDIA GPU** |
+| GTX 1080 Ti | 3.5 | 5.8 | 0.6 | - <sup>(Note 2) </sup> | - | 2.7 | - | - | 4.0 | 4.6 |
+| GTX TITAN X | 5.8 | 9.7 | 1.0 | - | - | 4.3 | - | - | 6.4 | 7.5 |
+| **AMD GPU** |
+| AMD Vega FE | 5.7 | 8.8 | 1.0 | - | - | 4.4 | - | - | 6.1 | 7.0 |
+| |
+
+* Note 1: Out of memory on this board.
+* Note 2: We didn't tune some small networks on GPU due to time limit. TVM can use its fallback mechanism to compile them but the performance is not guaranteed.
```

Comment:
... due to time contraints...
When profiling data is not available... TVM can use fallback code generation (but competitive performance is not guaranteed in this scenario).